src/Interpreters/Access/InterpreterShowGrantsQuery.cpp

```diff
@@ -135,15 +136,53 @@ QueryPipeline InterpreterShowGrantsQuery::executeImpl()
 
 std::vector<AccessEntityPtr> InterpreterShowGrantsQuery::getEntities() const
 {
-    const auto & show_query = query_ptr->as<ASTShowGrantsQuery &>();
+    const auto & access = getContext()->getAccess();
     const auto & access_control = getContext()->getAccessControl();
+
+    const auto & show_query = query_ptr->as<ASTShowGrantsQuery &>();
     auto ids = RolesOrUsersSet{*show_query.for_roles, access_control, getContext()->getUserID()}.getMatchingIDs(access_control);
 
+    bool throw_if_access_denied = !show_query.for_roles->all;
+    std::optional<bool> show_users_check_result;
```

Comment:
Not sure why we modify these variables inside lambdas and then return result of them, can't we just call lambda and return result ?
Also these lambdas are almost the same, maybe better to extract into some static method and call with `AccessType` argument ?

Reply:
I moved this logic to a separate helper class `CachedAccessChecking`, it seems it looks better now.
Perhaps later I'll use this helper class in some other places too...

Reply:
Agreed. Now much easier to read.